File: algorithms/medium/simple_bank_system.ex
defmodule Bank do
  use GenServer

  @spec init_(balance :: [integer]) :: any
  def init_(balance) do
    GenServer.start_link(__MODULE__, [], name: __MODULE__)
    balance = balance |> Enum.with_index(fn e, i -> {i + 1, e} end) |> Map.new()
    GenServer.call(__MODULE__, {:init, balance})
  end

  @spec transfer(account1 :: integer, account2 :: integer, money :: integer) :: boolean
  def transfer(account1, account2, money) do
    GenServer.call(__MODULE__, {:transfer, account1, account2, money})
  end

  @spec deposit(account :: integer, money :: integer) :: boolean
  def deposit(account, money) do
    GenServer.call(__MODULE__, {:deposit, account, money})
  end

  @spec withdraw(account :: integer, money :: integer) :: boolean
  def withdraw(account, money) do
    GenServer.call(__MODULE__, {:withdraw, account, money})
  end

  # Server
  def init(state) do
    {:ok, state}
  end

  def handle_call({:init, balance}, _from, _state) do
    {:reply, true, balance}
  end

  def handle_call({:transfer, account1, account2, money}, _from, balance) do
    cond do
      not Map.has_key?(balance, account1) ->
        {:reply, false, balance}

      not Map.has_key?(balance, account2) ->
        {:reply, false, balance}

      balance[account1] < money ->
        {:reply, false, balance}

      true ->
        balance = Map.update!(balance, account1, &(&1 - money))
        balance = Map.update!(balance, account2, &(&1 + money))
        {:reply, true, balance}
    end
  end

  def handle_call({:deposit, account, money}, _from, balance) do
    cond do
      not Map.has_key?(balance, account) ->
        {:reply, false, balance}

      true ->
        balance = Map.update!(balance, account, &(&1 + money))
        {:reply, true, balance}
    end
  end

  def handle_call({:withdraw, account, money}, _from, balance) do
    cond do
      not Map.has_key?(balance, account) ->
        {:reply, false, balance}

      balance[account] < money ->
        {:reply, false, balance}

      true ->
        balance = Map.update!(balance, account, &(&1 - money))
        {:reply, true, balance}
    end
  end
end

# Your functions will be called as such:
# Bank.init_(balance)
# param_1 = Bank.transfer(account1, account2, money)
# param_2 = Bank.deposit(account, money)
# param_3 = Bank.withdraw(account, money)

# Bank.init_ will be called before every test case, in which you can do some necessary initializations.

File: algorithms/medium/simple_bank_system.rs
struct Bank {
    balance: Vec<i64>,
}


/** 
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */
impl Bank {

    fn new(balance: Vec<i64>) -> Self {
        Self { balance }
    }
    
    fn transfer(&mut self, account1: i32, account2: i32, money: i64) -> bool {
        let account1 = account1 as usize;
        let account2 = account2 as usize;
        
        if account1 > self.balance.len() || account2 > self.balance.len() {
            return false;
        }
        
        if self.balance[account1-1] < money {
            return false;
        }
        
        self.balance[account1-1] -= money;
        self.balance[account2-1] += money;
        
        true
    }
    
    fn deposit(&mut self, account: i32, money: i64) -> bool {
        let account = account as usize;
        if account > self.balance.len() {
            return false;
        }
        
        self.balance[account-1] += money;
        true
    }
    
    fn withdraw(&mut self, account: i32, money: i64) -> bool {
        let account = account as usize;
        if account > self.balance.len() {
            return false;
        }
        
        if self.balance[account-1] < money {
            return false;
        }
        
        self.balance[account-1] -= money;
        true
    }
}

/**
 * Your Bank object will be instantiated and called as such:
 * let obj = Bank::new(balance);
 * let ret_1: bool = obj.transfer(account1, account2, money);
 * let ret_2: bool = obj.deposit(account, money);
 * let ret_3: bool = obj.withdraw(account, money);
 */
